Email editor – polish UI for template edit, swap and restore [MAILPOET-6336]

costasovo opened the pull request:
## Description

This PR removes the temporary template experiment panel and adds the controls to the proper places.


https://github.com/user-attachments/assets/de5e71fa-c580-470c-b164-6a6136520e28

## Code review notes

_N/A_

## QA notes

_N/A_

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-6336]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-6336]: https://mailpoet.atlassian.net/browse/MAILPOET-6336?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Preview

[Preview in WordPress Playground](https://account.mailpoet.com/playground/new/branch:email-editor/tempaltes-sidebar)

_The latest successful build from `email-editor/tempaltes-sidebar` will be used. If none is available, the link won't work._

costasovo committed:
Add a custom hook to determine whether we are in email or template mode

There are a couple of places where we determine the mode, but we use different approaches.
This commit adds a mechanism for determining the editor mode which can be easily reusable
in the whole app.
[MAILPOET-6336]

costasovo committed:
Replace logic for detection of editor mode by the useEditorMode hook

[MAILPOET-6336]

costasovo committed:
Rename mainSidebarEmailTab constant to mainSidebarDocumentTab

The email name was misleading because in the template mode we want to show Template.
[MAILPOET-6336]

costasovo committed:
Rename document tab in sidebar based on editor mode

[MAILPOET-6336]

costasovo committed:
Update text in template mode warning popup

[MAILPOET-6336]

costasovo committed:
Add info text to the template info panel in sidebar.

[MAILPOET-6336]

costasovo committed:
Add template actions menu toggle with dummy reset template action

[MAILPOET-6336]

costasovo committed:
Move revert template function from component file to store action

[MAILPOET-6336]

costasovo committed:
Add revert template confirmation modal

[MAILPOET-6336]

costasovo committed:
Extract template reset modal to an extra component

[MAILPOET-6336]

costasovo committed:
Add dropdown with edit and swap template buttons

[MAILPOET-6336]

costasovo committed:
Add edit template modal

[MAILPOET-6336

costasovo committed:
Add basic swap template functionality

[MAILPOET-6336]

costasovo committed:
Use current email content when previewing templates when swapping

[MAILPOET-6336]

costasovo committed:
Remove experimental templates panel

[MAILPOET-6336]

costasovo committed:
Update variant of the dropdown button to match the native editor

[MAILPOET-6336]

costasovo committed:
Hide command palette shortcut

[MAILPOET-6336]

costasovo committed:
Add basic acceptance test for template edit, swap and restore

[MAILPOET-6336]

triple0t commented on packages/js/email-editor/src/components/sidebar/email-type-info.tsx:
```diff
@@ -1,26 +1,125 @@
+							<Flex justify={ 'start' }>
+								<FlexItem className="editor-post-panel__row-label">
+									Template
+								</FlexItem>
```
Why not use the i18n function `__` here?

costasovo replied:
Nice catch! I added the translation function https://github.com/mailpoet/mailpoet/pull/5978/commits/739cb856322965693b1a7fb3709105a24a68efca

triple0t commented on packages/js/email-editor/src/components/header/index.scss:
```diff
@@ -39,3 +39,8 @@
+// We currently don't support command palette so we hide the shortcut
+.editor-document-bar__shortcut {
+  display: none;
+}
```
I can still see the command palette.

<img width="564" alt="Screenshot 2024-12-06 at 7 47 26 AM" src="https://github.com/user-attachments/assets/8529e539-890e-42b1-997d-89b7a54b5b3b">

costasovo replied:
Thank you! I fixed it by replacing the original commit. https://github.com/mailpoet/mailpoet/pull/5978/commits/223d8f6dc358baf88f884caa90ac7e6808755cc0

triple0t reviewed (commented):
Well done. 
This PR looks good, and I'm excited about it.

I have minor remarks; please check them below.

<img width="303" alt="Screenshot 2024-12-06 at 7 56 32 AM" src="https://github.com/user-attachments/assets/0d5ac419-1c37-4068-92c1-4e9b697875a9">

Can we update this toast content? It's currently showing `site updated` instead of `template updated`.

<img width="535" alt="Screenshot 2024-12-06 at 9 53 58 AM" src="https://github.com/user-attachments/assets/d1f64afe-6ff3-4fb3-ac16-bfc3ad465495">

Would it make sense to have a cancel button here?
It wasn't totally obvious that clicking outside the modal would exit it.


<img width="266" alt="Screenshot 2024-12-06 at 9 58 56 AM" src="https://github.com/user-attachments/assets/9e3a1c36-107c-4e14-9b01-ce9aacf820ac">

This button is very well hidden. Is there a reason why we placed it in such a non-descript place?


<img width="575" alt="Screenshot 2024-12-06 at 9 59 05 AM" src="https://github.com/user-attachments/assets/fe343b13-fad5-4c75-9db0-12c2f2bec19d">

I feel this warning isn't too clear. The wording suggests it would clear my current customization when, in actual fact, it removes "ANY and ALL" customization done on the template.

Can we update the wording to something like:
`WARNING: This will clear ANY and ALL template customization. All updates made to the template will be lost. Do you want to proceed?`

costasovo committed:
Add missing translation function to template info panel

[MAILPOET-6336]

costasovo committed:
Add the cancel button to modal when switching to template mode

[MAILPOET-6336]

costasovo committed:
Use more informative text in the reset template dialog

[MAILPOET-6336]

costasovo committed:
Add a workaround for replacing core notice Site updated

I considered doing this using a filter 'i18n.gettext', but
with this approach, we would just replace the text. The notice also
contained an action with a link to the site's homepage, so I chose this mechanism instead.
Let's see if there will be more use-cases. We can eventually move it somewhere into the store
and proxy all messages via our store.
[MAILPOET-6336]

costasovo commented:
@triple0t Thank you for the thorough review! Your remarks are great and will help to improve our app.

> Can we update this toast content? It's currently showing site updated instead of template updated.

Done in https://github.com/mailpoet/mailpoet/pull/5978/commits/9879ee4bc251c9eb7a6d4ca89e07c566a35664bc Note that this `Site updated` notice is in core used also for updating global styles. So for now, as styles are part of template data, we are fine, but in the future task where we introduce global styles, we will have to change it to something more suitable. Perhaps `Email design updated` could work. But that's just for the future.

> Would it make sense to have a cancel button here?
It wasn't totally obvious that clicking outside the modal would exit it.

The specification in the ticket explicitly mentions only a single button (point 3.f.I.2), but I agree that Cancel could be helpful. I just want to add that the native post editor doesn't show any dialog at all. I understood that the purpose of this dialog is to educate users about template mode a bit. I added the button in https://github.com/mailpoet/mailpoet/pull/5978/commits/80a825b8af6db1cb75412dd89d01bb00d7958670

> This button is very well hidden. Is there a reason why we placed it in such a non-descript place?

This is exactly the same place as in the site editor (see screenshot). I agree it is hidden, but at the same time, I think this action should be used rarely so I would keep it as it is for now.
![Screenshot 2024-12-06 at 10 41 16](https://github.com/user-attachments/assets/aae7b94a-eaca-4784-91d3-b8654c6022bc)

> I feel this warning isn't too clear. The wording suggests it would clear my current customization when, in actual fact, it removes "ANY and ALL" customization done on the template.

I think it is a good idea to make the message more informative. I used your message as was. I would like to ask @pavel-mailpoet here to give feedback on the change of the reset template warning to `This will clear ANY and ALL template customization. All updates made to the template will be lost. Do you want to proceed?` Is that ok?

pavel-mailpoet commented:
> I think it is a good idea to make the message more informative. I used your message as was. I would like to ask @pavel-mailpoet here to give feedback on the change of the reset template warning to This will clear ANY and ALL template customization. All updates made to the template will be lost. Do you want to proceed? Is that ok?

Sounds good, but do you know what the core editor says?

costasovo commented:
> Sounds good, but do you know what the core editor says?

@pavel-mailpoet The core editor says: "Reset to default and clear all customizations?" I think it does no harm to be more informative in this case. But when we switch to the WP's native editor, we will get the original message. I have no strong opinion here.

costasovo commented:
@triple0t I believe I addressed/answered all your remarks and the PR should be ready for the second review.

triple0t reviewed (approved):
LGTM. 
Thanks for making the update.

I see this PR is linked to another PR. I'll assign it back to you. Please feel free to merge when ready.
